File: Server/src/Server.Application/ViewModels/BookViewModel.cs
// ReSharper disable UnusedAutoPropertyAccessor.Global
#pragma warning disable CS8618

namespace Server.Application.ViewModels;

public class BookViewModel
{
    public Guid Id { get; set; }
    public string ISBN { get; set; }
    public string Title { get; set; }
    public int YearPublished { get; set; }
    public double? AverageRating { get; set; }
    public double? AverageCriticRating { get; set; }
    public string? Description { get; set; }
    public string Language { get; set; }
    public int? PageCount { get; set; }
    public string? CoverPictureUrl { get; set; }
    public PublisherViewModel Publisher { get; set; }
    public ICollection<AuthorViewModel> Authors { get; set; }
    public ICollection<BookCategoryViewModel> BookCategories { get; set; }
    public bool? DoesUserObserve { get; set; }
}

File: client/src/models/BookViewModel.ts
import { z } from 'zod';
import PublisherViewModel from './PublisherViewModel';
import AuthorViewModel from './author/AuthorViewModel';
import BookCategoryViewModel from './BookCategoryViewModel';
import ReviewViewModel from './reviews/ReviewViewModel';

const BookViewModel = z.object({
  id: z.string().uuid(),
  isbn: z.string().max(17),
  title: z.string(),
  yearPublished: z.number().int().positive(),
  description: z.string().nullable(),
  language: z.string(),
  pageCount: z.number().nullable(),
  coverPictureUrl: z.string().max(256).nullable(),
  averageRating: z.nullable(z.number()),
  averageCriticRating: z.nullable(z.number()),
  publisher: PublisherViewModel.nullable(),
  authors: z.array(AuthorViewModel),
  bookCategories: z.array(BookCategoryViewModel),
  doesUserObserve: z.nullable(z.boolean()),
  userReview: ReviewViewModel.nullable().optional(),
});

export default BookViewModel;
export type BookViewModelType = z.infer<typeof BookViewModel>;
